File: examples/nostr-basics/02-publish-ack/go/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/gorilla/websocket"
	nostr "github.com/nbd-wtf/go-nostr"
)

func main() {
	relay := "ws://127.0.0.1:3100"
	if len(os.Args) > 1 {
		relay = os.Args[1]
	}
	content := os.Getenv("NOSTR_CONTENT")
	if content == "" {
		content = "hello from the NIP-01 Go publisher"
	}

	secretKey := nostr.GeneratePrivateKey()
	publicKey, err := nostr.GetPublicKey(secretKey)
	if err != nil {
		panic(err)
	}
	event := nostr.Event{PubKey: publicKey, CreatedAt: nostr.Now(), Kind: 1, Content: content}
	if err := event.Sign(secretKey); err != nil {
		panic(err)
	}

	conn, _, err := websocket.DefaultDialer.Dial(relay, nil)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	if err := conn.WriteJSON([]any{"EVENT", event}); err != nil {
		panic(err)
	}
	fmt.Println("sent EVENT", event.ID)

	for {
		_, data, err := conn.ReadMessage()
		if err != nil {
			panic(err)
		}
		var envelope []json.RawMessage
		if err := json.Unmarshal(data, &envelope); err != nil || len(envelope) == 0 {
			continue
		}
		var label string
		if err := json.Unmarshal(envelope[0], &label); err != nil {
			continue
		}
		switch label {
		case "OK":
			if len(envelope) < 4 {
				continue
			}
			var id string
			var accepted bool
			var reason string
			_ = json.Unmarshal(envelope[1], &id)
			_ = json.Unmarshal(envelope[2], &accepted)
			_ = json.Unmarshal(envelope[3], &reason)
			if id == event.ID {
				fmt.Printf("relay OK: accepted=%t reason=%q\n", accepted, reason)
				return
			}
		case "NOTICE":
			fmt.Println("relay NOTICE:", string(envelope[1]))
		default:
			fmt.Println("relay", label, string(data))
		}
	}
}

File: examples/nostr-basics/02-publish-ack/rust/src/main.rs
use std::error::Error;

use futures_util::{SinkExt, StreamExt};
use nostr::{EventBuilder, Keys, Kind};
use serde_json::Value;
use tokio_tungstenite::{connect_async, tungstenite::Message};

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let relay = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "ws://127.0.0.1:3100".to_string());
    let content = std::env::var("NOSTR_CONTENT")
        .unwrap_or_else(|_| "hello from the NIP-01 Rust publisher".to_string());
    let keys = Keys::generate();
    let event = EventBuilder::new(Kind::TextNote, content).sign_with_keys(&keys)?;
    let event_id = event.id.to_hex();

    println!("connecting to {relay}");
    let (mut ws, _) = connect_async(&relay).await?;
    ws.send(Message::Text(
        serde_json::json!(["EVENT", event]).to_string().into(),
    ))
    .await?;
    println!("sent EVENT {event_id}");

    while let Some(message) = ws.next().await {
        let message = message?;
        let Message::Text(text) = message else {
            continue;
        };
        let value: Value = serde_json::from_str(text.as_ref())?;
        let Some(array) = value.as_array() else {
            continue;
        };
        match array.first().and_then(Value::as_str) {
            Some("OK") if array.get(1).and_then(Value::as_str) == Some(&event_id) => {
                let accepted = array.get(2).and_then(Value::as_bool).unwrap_or(false);
                let reason = array.get(3).and_then(Value::as_str).unwrap_or("");
                println!("relay OK: accepted={accepted} reason={reason:?}");
                break;
            }
            Some("NOTICE") => println!("relay NOTICE: {}", array.get(1).unwrap_or(&Value::Null)),
            Some(other) => println!("relay {other}: {value}"),
            None => {}
        }
    }

    Ok(())
}
